### This is a chunk of codes for RNA-seq

#####################################
setwd('~/Desktop/HMH/')
library(dplyr)
library(ggplot2)
library(DESeq2)
library(edgeR)
library(reshape)
### raw data were generated by featurecounts
### read raw counts and save in rds format
list.files('rds/RNA_seq_features/NIK', full.names = T)
feature_list <- list.files('rds/RNA_seq_features/NIK', full.names = T, 
                           pattern = 'matrix.txt')
feature_list
length(feature_list)
tmp.int <- read.table(feature_list[1],
                      header = T, row.names = 1)
colnames(tmp.int) <- paste0(substr(feature_list[1], 28,33),'.count')
tmp.df <- tmp.int
for(i in 2:length(feature_list)){
  sample <- substr(feature_list,28,33)[i]
  tmp <- read.table(feature_list[i],
                    header = T, col.names = c('gene',paste0(sample,'.count')))
  tmp.df <- cbind(tmp.df, tmp %>% select(contains('count')))
}

tmp.df %>% head()
tmp.df1 <- tmp.df %>% select(contains('s.count'))
tmp.df2 <- tmp.df %>% select(contains('t.count'))
tmp.df <- cbind(tmp.df1, tmp.df2)
tmp.df %>% head()
tmp.df[c('Tcf7','Ctla4'),]

write.csv(tmp.df, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.t.only.value.csv')

raw.counts <- tmp.df
raw.counts %>% head(30)
raw.counts[c('Tcf7','Icos','Ctla4'),]

raw.counts %>% dim()
raw.counts <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.c.only.value.csv', 
                       row.names = 1)
raw.counts %>% head()

### filtering 
table(rowSums(raw.counts) == 0)
raw.counts <- raw.counts[rowSums(raw.counts) != 0,]
raw.counts %>% dim()

summary(rowSums(raw.counts))
boxplot(rowSums(raw.counts)) ## outliers : 
hist((rowSums(raw.counts)), breaks = 1000)

raw.counts[rowSums(raw.counts) >= 1e7,] # 1e7
which(rowSums(raw.counts) >= 1e7)
raw.counts <- raw.counts[-which(rowSums(raw.counts) >= 1e7)[[1]],]
raw.counts %>% dim()
raw.counts %>% colnames()

raw.counts[rowSums(raw.counts) >= 5e6,] # Actb, Tmsb4x

write.csv(raw.counts, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.c.only.outlier.filtered.value.csv')
raw.counts <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.c.only.outlier.filtered.value.csv', 
                       row.names = 1)
raw.counts %>% head(30)

### filter 2 over 3 in any conditions
tmp <- raw.counts !=0
tmp %>% colnames()

colnames(tmp) <- c(paste0('ctr_s',1:3),
                   paste0('nik_s',1:3),
                   paste0('ctr_t',1:3),
                   paste0('ikk_t',1:3),
                   paste0('nik_t',1:3))

f1 <- tmp[tmp[,grep('ctr_s', colnames(tmp))] %>% rowSums() >=2,] %>% rownames()
f2 <- tmp[tmp[,grep('nik_s', colnames(tmp))] %>% rowSums() >=2,] %>% rownames()
f3 <- tmp[tmp[,grep('ctr_t', colnames(tmp))] %>% rowSums() >=2,] %>% rownames()
f4 <- tmp[tmp[,grep('ikk_t', colnames(tmp))] %>% rowSums() >=2,] %>% rownames()
f5 <- tmp[tmp[,grep('nik_t', colnames(tmp))] %>% rowSums() >=2,] %>% rownames()


union(f1,f2) %>% length()
union(union(f1,f2),f3) %>% length()
union(union(union(union(f1,f2),f3),f4),f5) %>% length()

raw.counts.filtered <- raw.counts[union(union(union(union(f1,f2),f3),f4),f5),]
raw.counts.filtered %>% head()
raw.counts.filtered %>% dim()
colnames(raw.counts.filtered) <- colnames(tmp)
write.csv(raw.counts.filtered, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.t.only.2over3.filtered.raw.value.csv')



raw.counts.filtered <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.t.only.2over3.filtered.raw.value.csv',
                                row.names = 1)
count.mtx <- raw.counts.filtered
count.mtx %>% head()

## input data : info sheet
info <- data.frame(matrix(nrow = ncol(count.mtx), ncol = 3))
colnames(info) <- c('sample', 'cell_type','condition')
info$sample <- substr(colnames(count.mtx),1,6)
info$cell_type <- substr(info$sample,5,5)
info$condition <- substr(info$sample,1,5)
info

###############################################
######## total dds : DESeq ####################
dds <- DESeqDataSetFromMatrix(count.mtx, info, ~condition)
dds <- DESeq(dds)

### vst for PCA ### 
vsd <- vst(dds,blind=TRUE)
plotPCA(vsd, intgroup="condition")

# to remove the dependence of the variance on the mean
#plotPCA(vsd, intgroup="mutation")
pcaplot <- plotPCA(vsd, intgroup="condition", return=T)  #using the DESEQ2 plotPCA fxn we can
### drawing plot
theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
               panel.background = element_blank(), 
               axis.line = element_line(colour = "black"))
ggplot(pcaplot, aes(PC1,PC2, shape=condition)) + geom_point(size=3, alpha=0.8) + 
  theme + xlab('PC1:62% variance') + ylab('PC2:24% variance')









###### gene range information from TxDb database #####
###### gene range information needed for fpkm calculation
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicFeatures)

allGenes <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
allGenes_ranges <- data.frame(allGenes@ranges)

## add ENTREZID column from name info
allGenes_ranges$ENTREZID <- allGenes_ranges$names
## genes in count matrix
## convert genes into Entrezid
count.mtx <- raw.counts.filtered
geneset <- rownames(count.mtx) ## gene list
library(clusterProfiler)
genes_to_convert <- clusterProfiler::bitr(geneset, fromType = "SYMBOL", 
                                          toType = c("ENSEMBL","ENTREZID"), 
                                          OrgDb = "org.Mm.eg.db")
## remove duplicated genes (Symbol)
genes_to_convert <- genes_to_convert[!duplicated(genes_to_convert$SYMBOL),]
rownames(genes_to_convert) <- genes_to_convert$SYMBOL

## subset out the selected genes range information

selected_genes <- allGenes_ranges[allGenes_ranges$names %in% genes_to_convert$ENTREZID,]
selected_genes %>% dim()
## two inputs
## genes to convert : symbol, ENTREZID
## selected genes : ENTREZID, location, width 
## join columns by ENTREZID
gene_info <- left_join(genes_to_convert, selected_genes, by='ENTREZID')
gene_info %>% dim()

######### FPKM calculation ##########

fpkm <- RNAAgeCalc::count2FPKM(count.mtx[gene_info$SYMBOL,],
                               genelength = gene_info$width,idtype = "SYMBOL")
fpkm %>% head()

write.csv(fpkm, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.t.only.2over3.filtered.fpkm.csv')


fpkm %>% head()
gene <- 'Tcf7'
input.data <- fpkm
df <- input.data[gene,] %>% melt() %>% data.frame()
df$condition <- substr(df$variable,1,5)
df2 <- df %>% group_by(condition) %>% summarize(avg=mean(value), sd=sd(value))
ggplot(df2, aes(condition,avg)) + geom_col()+ geom_errorbar(aes(ymin=avg-sd, ymax=avg+sd))


gene.exp.func <- function(gene,input.data=fpkm){
  gene <- as.character(gene)
  if(gene %in% rownames(input.data)){
    df <- input.data[gene,] %>% melt() %>% data.frame()
    df$condition <- substr(df$variable,1,5)
    df2 <- df %>% group_by(condition) %>% summarize(avg=mean(value), sd=sd(value))
    p <- ggplot(df2, aes(condition,avg)) + geom_col()+ geom_errorbar(aes(ymin=avg-sd, ymax=avg+sd)) + 
      ylab(gene) + theme_classic()
    print(p)
  }
  else{
    message('no gene in this dataset')
  }
}



fpkm <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.import.feature.matrix.raw.t.only.2over3.filtered.fpkm.csv',
                 row.names = 1)
fpkm %>% head()
fpkm %>% dim()

is.na(rowSums(fpkm)) %>% count()
na.genes <- is.na(rowSums(fpkm)) %>% which() %>% names()
is.na(rowSums(fpkm)) %>% names()
fpkm <- fpkm[!is.na(fpkm$ctr_s1),]
write.csv(fpkm, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.t_only.na_filtered.fpkm.csv')
fpkm <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.t_only.na_filtered.fpkm.csv',
                 row.names = 1)
fpkm %>% head()

df <- fpkm[, grep('ctr_s', colnames(fpkm))]
df %>% head()
rowMeans(df) %>% mean()
avg.mtx <- data.frame(matrix(ncol = 3, nrow = nrow(df)))
avg.mtx[,1] <- rowMeans(df) %>% mean()
avg.mtx[,2] <- rowMeans(df) %>% mean()
avg.mtx[,3] <- rowMeans(df) %>% mean()
avg.mtx %>% dim()
avg.mtx %>% head()
df %>% dim()
df.avg <- df/avg.mtx 
df.avg %>% head()
df[1,]/avg.mtx[1,]
df[2,]/avg.mtx[2,]





##########################################

df <- fpkm[, grep('ctr_s', colnames(fpkm))]
avg.mtx <- data.frame(matrix(ncol = 3, nrow = nrow(df)))
avg.mtx[,1] <- rowMeans(df) %>% mean()
avg.mtx[,2] <- rowMeans(df) %>% mean()
avg.mtx[,3] <- rowMeans(df) %>% mean()
df.avg <- df/avg.mtx 

for(i in c('ctr_t','nik_s','nik_t','ikk_t')){
  df <- fpkm[, grep(i, colnames(fpkm))]
  avg.mtx <- data.frame(matrix(ncol = 3, nrow = nrow(df)))
  avg.mtx[,1] <- rowMeans(df) %>% mean()
  avg.mtx[,2] <- rowMeans(df) %>% mean()
  avg.mtx[,3] <- rowMeans(df) %>% mean()
  df.avg.tmp <- df/avg.mtx
  df.avg <- cbind(df.avg, df.avg.tmp)
}
df.avg %>% head()
fpkm.norm <- df.avg

gene <- 'Cxcr5'
p1 <- gene.exp.func(gene)
p2 <- gene.exp.func(gene, input.data = fpkm.norm)
cowplot::plot_grid(p1,p2)


write.csv(fpkm.norm, 'rds/RNA_seq/NIK_t/RNA_seq_NIK.t_only.na_filtered.norm.fpkm.csv')
fpkm.norm <- read.csv('rds/RNA_seq/NIK_t/RNA_seq_NIK.t_only.na_filtered.norm.fpkm.csv',
                 row.names = 1)
fpkm.norm %>% head()

fpkm.norm['Tcf7',]
